core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingInterceptor.java

```diff
@@ -129,20 +132,20 @@ private boolean performPartitionCheck(InvocationContext ctx, LocalFlagAffectedCo
    }
 
    private CompletableFuture<Void> handleDataReadCommand(InvocationContext ctx, DataCommand command) {
-      Object key = command.getKey();
       return ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
+         DataCommand dataCommand = (DataCommand) rCommand;
          if (throwable != null) {
-            if (throwable instanceof RpcException && performPartitionCheck(rCtx, ((DataCommand) rCommand))) {
+            if (throwable instanceof RpcException && performPartitionCheck(rCtx, dataCommand)) {
                // We must have received an AvailabilityException from one of the owners.
                // There is no way to verify the cause here, but there isn't any other way to get an invalid
                // get response.
-               throw log.degradedModeKeyUnavailable(key);
+               throw log.degradedModeKeyUnavailable(dataCommand.getKey());
```

Comment:
One-way communication is not really tested; it's behaves a malicious node and I am not sure if the code is bulletproof to that.
Let's deal with this rather in a separate PR, with proper tests.